In [1]:
import pandas as pd

df = pd.read_csv("messy_sales.csv")

print(df.shape)
display(df.head(10))
print(df.dtypes)

(300, 6)


,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


order_id      int64
date            str
product         str
price       float64
qty           int64
zip           int64
dtype: object


The date format has three diffrent shape,on the price colune there are a missing price,the zip code missing a same dight number

In [2]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

In [3]:
fill_value = df["price"].median()

print(fill_value)

37.53


In [4]:
df["price"] = df["price"].fillna(fill_value)

In [5]:
df.isna().sum()

order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

In [6]:
print(df.duplicated().sum())

8


In [7]:
df = df.drop_duplicates()

In [8]:
print(df.shape)

(292, 6)


In [9]:
df["zip"] = df["zip"].astype(str).str.zfill(5)

In [10]:
df["zip"].head(10)

0    60614
1    30303
2    10001
3    98101
4    90405
5    30303
6    02134
7    90405
8    60614
9    90210
Name: zip, dtype: str

In [11]:
df["date"] = pd.to_datetime(df["date"], format="mixed")

In [12]:
print(df.dtypes)

order_id             int64
date        datetime64[us]
product                str
price              float64
qty                  int64
zip                    str
dtype: object


In [13]:
df.isna().sum()


order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

In [14]:
print(df.shape)
print(df.dtypes)
print(df.isna().sum())

(292, 6)
order_id             int64
date        datetime64[us]
product                str
price              float64
qty                  int64
zip                    str
dtype: object
order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64


In [15]:
df[df["qty"] < 0]

,order_id,date,product,price,qty,zip
202,1140,2026-04-18,webcam,38.69,-5,98101
262,1233,2026-05-09,desk lamp,22.64,-4,10001
297,1025,2026-04-27,keyboard,47.30,-2,02116


In [16]:
df.loc[df["qty"] < 0, "qty"] = df.loc[df["qty"] < 0, "qty"].abs()

In [17]:
df[df["qty"] < 0]

,order_id,date,product,price,qty,zip


I consider the negative quantities as a typing mistake casue it is an order quantity with a price and address which reflet it is something required by a real demande 

In [18]:
print(df.shape)
df.head()

(292, 6)


,order_id,date,product,price,qty,zip
0,1254,2026-04-06,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,2026-05-06,mug,5.50,1,10001
3,1066,2026-05-30,notebook,37.53,1,98101
4,1114,2026-04-06,webcam,45.59,4,90405


In [19]:
print(df.shape)

(292, 6)


In [20]:
df.to_csv("sales_clean.csv", index=False)

In [21]:
import os

print(os.listdir())

['.ipynb_checkpoints', 'module3_lab.ipynb', 'messy_sales.csv', 'sales_clean.csv']


 Cleaning Log

Loaded 300 rows and 6 columns from messy_sales.csv file.
Filled  the missing prices using the Tutorial 2 pattern ( 12 missing price values using the median value of (35.75 ) because the median is less affected by skewed prices.)

Removed 8 duplicate rows, leaving 292 rows.
 Converted ZIP codes from numbers to text and restored five-character ZIP codes, including leading zeros.
Converted the date column from text to a datetime format.

 Found 3 negative quantity values. I treated them as typing mistakes and changed them to positive quantities.
 The cleaned dataset is saved as `sales_clean.csv`.

The risk of a wrong cleaning decision could change the results of an analysis and lead a business to make a decision based on inaccurate information.which I point ( how i treat the negative quantity) it might be anyother point of view.